packages/kit/src/runtime/server/page/actions.js

```diff
@@ -280,6 +280,13 @@ function try_deserialize(data, fn, route_id) {
 		// If we're here, the data could not be serialized with devalue
 		const error = /** @type {any} */ (e);
 
+		// if someone tries to use `json()` in their action
+		if (data instanceof Response) {
+			throw new Error(
+				`Data returned from action inside ${route_id} is not serializable. Form actions need to return plain objects or fail(). e.g. return { success: true } or return fail(400, { message: "invalid" });`
+			);
+		}
+
 		if ('path' in error) {
```

Comment:
I know this is existing code, but could you add a comment here explaining when `path` is or isn't defined? I'm wondering when we hit this branch vs falling through to rethrowing the original error

Reply:
Sure! If I had to guess, it's when the error is thrown by devalue itself.